### PR TITLE
Feat: Add file tree, add include and exclude prefix for auto ingestion

### PR DESCRIPTION
**Pull request by StrongMonkey**
This PR add file tree to show knowledge files. Demo is here https://www.loom.com/share/0e707405c1194ea1a9a1ffd595347a22 

It also adds a backend piece of functionality to include/exclude folder prefix so that user can add a folder in UI to automatically ingest files inside the folder.

### CODE REVIEW

File: pkg/controller/handlers/knowledgefile/knowledgefile.go
Comment (njhale) [inline pkg/controller/handlers/knowledgefile/knowledgefile.go]:
```suggestion
func hasAnyPrefix(filePath string,  prefixes []string) bool {
	for _, prefix := range prefixes {
		if strings.HasPrefix(filePath, prefix) {
			return true
		}
	}

	return false
}
```

File: ui/admin/app/components/knowledge/FileTree.tsx
Comment (ryanhopperlowe) [inline ui/admin/app/components/knowledge/FileTree.tsx]: use `visible`/`invisible` instead of `hidden`/`block`

This will prevent the row from resizing when it adds the icon button to the row

```suggestion
                                        className="justify-start items-center group ml-2 invisible group-hover:visible"
```

File: ui/admin/app/components/knowledge/FileTree.tsx
Comment (ryanhopperlowe) [inline ui/admin/app/components/knowledge/FileTree.tsx]: can you remove `hover:cursor-pointer`? clicking the div doesn't actually do anything

File: pkg/controller/handlers/knowledgefile/knowledgefile.go
Comment (njhale) [inline pkg/controller/handlers/knowledgefile/knowledgefile.go]:
```suggestion
	if file.Spec.Approved == nil {
	    var approved *bool
	    
	    switch {
	    case source.Spec.Manifest.AutoApprove != nil && *source.Spec.Manifest.AutoApprove:
	        approved = typed.Pointer(true)
	    case matchExclude:
	        approved = typed.Pointer(false)
	    case matchInclude:
	        approved = typed.Pointer(true)
	    }
	    
		if approved != nil {
		    file.Spec.Approved = approved
		    if err := req.Client.Update(req.Ctx, file); err != nil {
		        return err
		    }
		}
	}
```

**Review by njhale — CHANGES_REQUESTED** [review]

File: ui/admin/app/components/knowledge/FileTree.tsx
Comment (StrongMonkey) [reply ui/admin/app/components/knowledge/FileTree.tsx]: Actually it will open and close the folder tree that's why I added this.

Comment (ryanhopperlowe) [reply ui/admin/app/components/knowledge/FileTree.tsx]: I see, can you remove it when the row has no children then? I'm still seeing it for leaf nodes

Comment (StrongMonkey) [reply ui/admin/app/components/knowledge/FileTree.tsx]: Oh sure 👍

File: pkg/controller/handlers/knowledgefile/knowledgefile.go
Comment (njhale) [inline pkg/controller/handlers/knowledgefile/knowledgefile.go]: The code from before you addressed my comment didn't perform an update if the value of `file.Spec.Approved` was still `nil` after checking all the approve/deny conditions; in other words, it didn't update when "no decision" about approval was reached.

This updated code __always__ performs an update, even when there's no change to `file.Spec.Approved`.

Was this change in behavior intentional?

**Review by njhale — CHANGES_REQUESTED** [review]

**Review by njhale — APPROVED** [review]

File: apiclient/types/knowledgesource.go
Comment (thedadams) [inline apiclient/types/knowledgesource.go]: Do we need this both here and `KnowledgeSourceManifest`? Seems like just in `KnowledgeSourceManifest` is enough.

File: pkg/api/handlers/knowledgesource.go
Comment (thedadams) [inline pkg/api/handlers/knowledgesource.go]: Can be removed if we use the fields from `KnowledgeSourceManifest`.

File: pkg/controller/handlers/knowledgefile/knowledgefile.go
Comment (thedadams) [inline pkg/controller/handlers/knowledgefile/knowledgefile.go]: nit: only do the processing that's necessary.
```suggestion
		switch {
		case source.Spec.Manifest.AutoApprove != nil && *source.Spec.Manifest.AutoApprove:
			file.Spec.Approved = typed.Pointer(true)
		case isFileMatchPrefixPattern(file.Spec.FileName, source.Spec.Manifest.FilePathPrefixInclude):
			file.Spec.Approved = typed.Pointer(false)
		case isFileMatchPrefixPattern(file.Spec.FileName, source.Spec.Manifest.FilePathPrefixInclude):
			file.Spec.Approved = typed.Pointer(true)
		}
```

**Review by thedadams — CHANGES_REQUESTED** [review]

**Review by ryanhopperlowe — APPROVED** [review]

**Comment by ryanhopperlowe** [comment]
Screenshot 2024-11-21 at 9 34 30 AM

Noticing that the add button is getting cutoff when the file name is too long. I think we need to truncate the text and add a tooltip to it

**Review by ryanhopperlowe — CHANGES_REQUESTED** [review]
discovered above issue

File: pkg/api/handlers/knowledgesource.go
Comment (StrongMonkey) [reply pkg/api/handlers/knowledgesource.go]: yeah, right. I will drop that

**Comment by ryanhopperlowe** [comment]
> @ryanhopperlowe good catch, should be fixed 🙏 image

Beautiful 🙌

**Review by ryanhopperlowe — APPROVED** [review]

**Review by iwilltry42 — APPROVED** [review]

**Review by thedadams — APPROVED** [review]

**Pull request merged**
